### DISCUSSION TITLE
Recurrent errors during split/merge tasks

### DISCUSSION THREAD

**Discussion by carsonpurnell** [nextpyp/.github#14] [Q&A] [answered]
I've been unable to run the tomography tutorial workflow, and getting errors during the split/merge. Logs are attached - err1 and err2 are from the GUI for the splits that at least provided error logs, the others are the files those logs seem to be pointing at. Let me know if more are needed.

This may or may not be related to a prior issue where those jobs wouldn't run at all due to apparently insufficient memory allocation - giving them all the RAM and only one thread appears to have solved that, but might be contributing to this? Defaults that work for the tutorial workflow would be nice to have.

[err1.txt](https://github.com/nextpyp/.github/files/13478932/err1.txt)
[err2.txt](https://github.com/nextpyp/.github/files/13478933/err2.txt)
[TS_01_csp.log](https://github.com/nextpyp/.github/files/13478934/TS_01_csp.log)
[TS_03_csp.log](https://github.com/nextpyp/.github/files/13478935/TS_03_csp.log)
[TS_43_csp.log](https://github.com/nextpyp/.github/files/13478936/TS_43_csp.log)
[TS_45_csp.log](https://github.com/nextpyp/.github/files/13478937/TS_45_csp.log)
[TS_54_csp.log](https://github.com/nextpyp/.github/files/13478938/TS_54_csp.log)

**Comment by abartesaghi** [comment]
We tracked the error you are getting to an issue with the ratio of RAM to threads being too high (64 GB/1 thread in your case). We will fix that in the next release, but you should still be able to run refinement using a more realistic ratio of 4-8GB RAM per thread. For example, if you have 64GB of RAM, you could use 8-16 `Threads per task` (depending on how many cores you have in your system). It is difficult to provide defaults because the optimal settings depend on the specific hardware configuration, but for refinement in general, you should use 4-8GB RAM per thread and as many `Threads per task` as your system allows. For reference, we used `Threads per task` = 124 and `Memory per task` = 720 GB (~6 GB/thread) for the [tutorials](https://nextpyp.app/files/pyp/latest/docs/tutorials.html).

**Comment by carsonpurnell** [comment]
Changing thread counts seems to get past that error (using 8 threads 64G to undershoot available RAM), but is now softlocking the workstation (despite undershooting RAM and CPUs by a significant margin). the UI does not show any error or generate any error log locations, this is the only file that shows anything that looks relevant after a restart.

[out-eoe1AbNtxcegQxWi.2.log](https://github.com/nextpyp/.github/files/13516055/out-eoe1AbNtxcegQxWi.2.log)

**Reply by abartesaghi** [reply]
Assuming you are running nextpyp in standalone mode (i.e., not using a SLURM cluster), this error could be caused by the way resources are managed in this mode. If you are running the tomography tutorial using `Threads per task = 8` and `Memory per task = 64`, nextpyp will need to run 5 such tasks (one for each tilt-series). For example, if your system has 17 cores, nextpyp will be able to fit up to (17-1)/8=2 tasks simultaneously and will assume you have the necessary 64x2=128GB of RAM available. Since the program does not track how much memory is available or is utilized (the way slurm does, for example), you could very easily overload the server if you don't have enough resources. How many total cores and memory does your system have?

**Reply by carsonpurnell** [reply]
After running in different combinations knowing how it uses tasks in the math, it still fails while far under any hardware limits. 5 tasks with 8 threads/8G is well under the system's 64 cores and 128G RAM. Error log (appears to be identical for all 5 split job in the refinement step)

[8thread8ramerr.txt](https://github.com/nextpyp/.github/files/13570235/8thread8ramerr.txt)

**Reply by abartesaghi** [reply, marked as answer]
Right, 1G/thread will not be enough memory for nextPYP to run, we recommend using between 4-8GB/thread as mentioned above. With the resources you have, you may get away using `Threads per task = 8` and `Memory per task = 16` (at least for the early stages of refinement). For the merge task, you may use `Threads (merge task) = 6` and `Memory (merge task) = 24`. Note that since you will be using only 8 threads per task, your running times will be significantly slower than the numbers we report in the tutorial.
